### ISSUE TITLE
`smooth` keyword in `bootstrap` should be configurable and have different default

### ISSUE THREAD

**Issue by HDembinski** [dsaxton/resample#7]
Currently, it is just true or false. No information is given in the docs how broad the Gaussian noise is, I see in the code that you use a sigma = 1/sqrt(N), how is this motivated? It seems ad hoc. I think the noise sigma should be proportional to the sample std. deviation, otherwise it is inconsistent. If the original distribution is very narrow, then also the noise must be narrow and vice versa.

I think there is no one answer how the smoothing should be done. In that case, the sigma should be configurable. I am not a fan of smoothing, btw, it goes against the bootstrap principles in my view.

To summarize, I think the new default noise width should be (std.deviation of original sample) / sqrt(N - 1) (aka error on the mean), and users should be allowed to pass in any float which is then used as the noise width.

**Comment by dsaxton** [comment]
I agree it is somewhat ad hoc; the original idea was to apply more smoothing the smaller the sample (as smaller samples are more "coarse"). I guess it's a question of how opinionated we want to be, but I also wouldn't be opposed to removing this option, as it either has little effect (small variance) or significantly changes the statistical properties of the sample (large variance), which partially invalidates results.

If we did keep it then yes it probably should somehow be proportional to the underlying variance.

**Comment by HDembinski** [comment]
I don't know how many users you already have, whether taking interface away at this point is still ok without hurting too many. The soft option is to deprecate things (I do that excessively in iminuit right now to clean up the interface design left by previous maintainers), keep the keyword in the code, but remove them from the documentation, and raise warnings when people are still using it.

While I personally don't see the point of smoothing, there are research papers about that, so maybe it makes sense to keep it.

**Comment by dsaxton** [comment]
As far as I can tell there aren't a ton of users (I see about 13K downloads over the past 1.5 years or so on pepy.tech) and I doubt it's used "in production" much so I'm actually not too worried about breaking changes.

In regard to smoothing, you could argue that it's somewhat redundant given that parametric bootstrapping is already possible (instead of adding Gaussian noise you could instead just use the Gaussian parametric bootstrap and then at least preserve the scale of the distribution). I'm slightly biased in favor of removing unnecessary details / features, but I suppose I don't feel too strongly either way.

**Comment by HDembinski** [comment]
> In regard to smoothing, you could argue that it's somewhat redundant given that parametric bootstrapping is already possible (instead of adding Gaussian noise you could instead just use the Gaussian parametric bootstrap and then at least preserve the scale of the distribution).

I agree. Smoothing changes the empirical CDF of the data. Doing a parametric bootstrap is better if one really has so few that data points that smoothing seems necessary. When the amount of data is large, no smoothing is needed anyway (at least I am not aware of a counter-example).

> I'm slightly biased in favor of removing unnecessary details / features, but I suppose I don't feel too strongly either way.

Then let's do that and have a clean starting point.
